File: python/examples/memory_ops.py
#!/usr/bin/env python3
# Copyright (c) 2024-2026 Elias Bachaalany
# SPDX-License-Identifier: MPL-2.0
#
# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this
# file, You can obtain one at https://mozilla.org/MPL/2.0/.
#
# memory_ops: Memory block inspection, byte reads, writes, and batch patches.
#
# Usage: python memory_ops.py [host_url]
#
# Defaults: http://127.0.0.1:18080, expects a program already open in Ghidra.

import sys

import libghidra as ghidra
from libghidra.models import BytePatch


def hex_dump(data: bytes, base_addr: int, width: int = 16) -> str:
    """Format raw bytes as a classic hex dump."""
    lines = []
    for i in range(0, len(data), width):
        chunk = data[i : i + width]
        hex_part = " ".join(f"{b:02x}" for b in chunk)
        ascii_part = "".join(chr(b) if 32 <= b < 127 else "." for b in chunk)
        lines.append(f"  0x{base_addr + i:08x}  {hex_part:<{width * 3}}  {ascii_part}")
    return "\n".join(lines)


def main() -> None:
    url = sys.argv[1] if len(sys.argv) >= 2 else "http://127.0.0.1:18080"
    client = ghidra.connect(url)

    # 1. List all memory blocks
    try:
        resp = client.list_memory_blocks()
    except ghidra.GhidraError as e:
        print(f"list_memory_blocks failed: {e}", file=sys.stderr)
        sys.exit(1)

    blocks = resp.blocks
    print(f"Memory blocks ({len(blocks)}):")
    for blk in blocks:
        perms = ("R" if blk.is_read else "-") + ("W" if blk.is_write else "-") + ("X" if blk.is_execute else "-")
        print(f"  {blk.name:<16} 0x{blk.start_address:08x}..0x{blk.end_address:08x}  {blk.size:>8} bytes  {perms}")

    if not blocks:
        print("No memory blocks found -- is a program open?", file=sys.stderr)
        sys.exit(1)

    # 2. Read 64 bytes from the start of the first initialized block
    target = next((b for b in blocks if b.is_initialized), blocks[0])
    read_len = min(64, target.size)
    addr = target.start_address

    try:
        read_resp = client.read_bytes(addr, read_len)
    except ghidra.GhidraError as e:
        print(f"read_bytes failed: {e}", file=sys.stderr)
        sys.exit(1)

    print(f"\nRead {len(read_resp.data)} bytes from '{target.name}' at 0x{addr:08x}:")
    print(hex_dump(read_resp.data, addr))

    # Save original bytes for restoration
    original = read_resp.data[:4]

    # 3. Write 4 bytes at the same address
    patch_data = bytes([0xDE, 0xAD, 0xBE, 0xEF])
    try:
        write_resp = client.write_bytes(addr, patch_data)
        print(f"\nWrote {write_resp.bytes_written} bytes at 0x{addr:08x}")
    except ghidra.GhidraError as e:
        print(f"write_bytes failed: {e}", file=sys.stderr)
        sys.exit(1)

    # Verify the write
    verify = client.read_bytes(addr, 4)
    print(f"Verify: {verify.data.hex()}")

    # 4. Batch-patch two locations at once
    patches = [
        BytePatch(address=addr, data=original),
        BytePatch(address=addr + 2, data=bytes([0xCA, 0xFE])),
    ]
    try:
        batch_resp = client.patch_bytes_batch(patches)
        print(f"\nBatch patch: {batch_resp.patch_count} patches, {batch_resp.bytes_written} bytes written")
    except ghidra.GhidraError as e:
        print(f"patch_bytes_batch failed: {e}", file=sys.stderr)
        sys.exit(1)

    # Final read to show result
    final = client.read_bytes(addr, 8)
    print(f"Final state at 0x{addr:08x}: {final.data.hex()}")

    # 5. Restore original bytes
    client.write_bytes(addr, original)
    restored = client.read_bytes(addr, 4)
    print(f"Restored: {restored.data.hex()}")

    print("\nDone.")


if __name__ == "__main__":
    main()

File: rust/examples/memory_ops.rs
// Copyright (c) 2024-2026 Elias Bachaalany
// SPDX-License-Identifier: MPL-2.0
//
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at https://mozilla.org/MPL/2.0/.
//
// memory_ops: Memory block inspection, byte reads, writes, and batch patching.
//
// Usage: memory_ops [host_url]

use libghidra as ghidra;

fn main() {
    let args: Vec<String> = std::env::args().collect();
    let url = args.get(1).map_or("http://127.0.0.1:18080", |s| s.as_str());
    let client = ghidra::connect(url);

    // 1. List memory blocks
    let blocks = match client.list_memory_blocks(50, 0) {
        Ok(b) => b,
        Err(e) => {
            eprintln!("ListMemoryBlocks failed (is a program open?): {e}");
            std::process::exit(1);
        }
    };

    println!("Memory blocks ({}):", blocks.blocks.len());
    for blk in &blocks.blocks {
        println!(
            "  {:<16} 0x{:08x}..0x{:08x}  size={:<8} R={} W={} X={}",
            blk.name,
            blk.start_address,
            blk.end_address,
            blk.size,
            blk.is_read as u8,
            blk.is_write as u8,
            blk.is_execute as u8,
        );
    }

    // 2. Read the first 64 bytes from the first block
    let first = match blocks.blocks.first() {
        Some(b) => b,
        None => {
            eprintln!("No memory blocks found.");
            std::process::exit(1);
        }
    };

    let read_len: u32 = std::cmp::min(64, first.size as u32);
    let read = client
        .read_bytes(first.start_address, read_len)
        .unwrap_or_else(|e| {
            eprintln!("ReadBytes failed: {e}");
            std::process::exit(1);
        });

    println!(
        "\nFirst {} bytes of '{}' at 0x{:08x}:",
        read.data.len(),
        first.name,
        first.start_address
    );
    hex_dump(&read.data, first.start_address);

    // 3. Write a small marker (4 bytes) at the block start
    let marker: Vec<u8> = vec![0xDE, 0xAD, 0xBE, 0xEF];
    let write = client
        .write_bytes(first.start_address, &marker)
        .unwrap_or_else(|e| {
            eprintln!("WriteBytes failed: {e}");
            std::process::exit(1);
        });
    println!(
        "\nWriteBytes: {} bytes written at 0x{:08x}",
        write.bytes_written, first.start_address
    );

    // 4. Verify the write by reading back
    let verify = client.read_bytes(first.start_address, 4).unwrap();
    println!("Read-back after write:");
    hex_dump(&verify.data, first.start_address);

    // 5. Batch-patch two locations: start+8 and start+16
    let patches = vec![
        ghidra::BytePatch {
            address: first.start_address + 8,
            data: vec![0xCA, 0xFE],
        },
        ghidra::BytePatch {
            address: first.start_address + 16,
            data: vec![0xBA, 0xBE],
        },
    ];
    let batch = client.patch_bytes_batch(&patches).unwrap_or_else(|e| {
        eprintln!("PatchBytesBatch failed: {e}");
        std::process::exit(1);
    });
    println!(
        "\nPatchBytesBatch: {} patches, {} bytes written",
        batch.patch_count, batch.bytes_written
    );

    // 6. Final read to see all modifications
    let final_read = client.read_bytes(first.start_address, 32).unwrap();
    println!("\nFinal state (first 32 bytes):");
    hex_dump(&final_read.data, first.start_address);
}

/// Print a classic hex dump: offset, hex bytes, ASCII.
fn hex_dump(data: &[u8], base: u64) {
    for (i, chunk) in data.chunks(16).enumerate() {
        let addr = base + (i * 16) as u64;
        print!("  {:08x}  ", addr);
        for (j, &b) in chunk.iter().enumerate() {
            if j == 8 {
                print!(" ");
            }
            print!("{:02x} ", b);
        }
        // Pad if less than 16 bytes
        for j in chunk.len()..16 {
            if j == 8 {
                print!(" ");
            }
            print!("   ");
        }
        print!(" |");
        for &b in chunk {
            let c = if b.is_ascii_graphic() || b == b' ' {
                b as char
            } else {
                '.'
            };
            print!("{c}");
        }
        println!("|");
    }
}
